crates/common/consensus/lean/src/lib.rs

```diff
@@ -1,4 +1,213 @@
-pub mod attestation;
 pub mod block;
+pub mod staker;
 pub mod state;
-pub mod validator;
+pub mod vote;
+
+use std::collections::HashMap;
+
+use alloy_primitives::B256;
+use serde::{Deserialize, Serialize};
+use ssz_types::{
+    VariableList,
+    typenum::{
+        U16777216, // 2**24
+    },
+};
+
+use crate::{block::Block, state::LeanState, vote::Vote};
+
+pub type Hash = B256;
+
+pub const ZERO_HASH: Hash = Hash::ZERO;
+pub const SLOT_DURATION: usize = 12;
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum QueueItem {
+    BlockItem(Block),
+    VoteItem(Vote),
+}
+
+// We allow justification of slots either <= 5 or a perfect square or oblong after
+// the latest finalized slot. This gives us a backoff technique and ensures
+// finality keeps progressing even under high latency
+pub fn is_justifiable_slot(finalized_slot: &usize, candidate_slot: &usize) -> bool {
+    assert!(
+        candidate_slot >= finalized_slot,
+        "Candidate slot ({candidate_slot}) is less than finalized slot ({finalized_slot})"
+    );
+
+    let delta = candidate_slot - finalized_slot;
+
+    delta <= 5
+    || (delta as f64).sqrt().fract() == 0.0 // any x^2
+    || (delta as f64 + 0.25).sqrt() % 1.0 == 0.5 // any x^2+x
+}
+
+// Given a state, output the new state after processing that block
+pub fn process_block(pre_state: &LeanState, block: &Block) -> LeanState {
+    let mut state = pre_state.clone();
+
+    // Track historical blocks in the state
+    // TODO: proper error handlings
+    let _ = state.historical_block_hashes.push(block.parent);
+    let _ = state.justified_slots.push(false);
+
+    while state.historical_block_hashes.len() < block.slot {
+        // TODO: proper error handlings
+        let _ = state.justified_slots.push(false);
+        let _ = state.historical_block_hashes.push(None);
+    }
+
+    // Process votes
+    for vote in &block.votes {
+        // Ignore votes whose source is not already justified,
+        // or whose target is not in the history, or whose target is not a
+        // valid justifiable slot
+        if !state.justified_slots[vote.data.source_slot]
+            || Some(vote.data.source) != state.historical_block_hashes[vote.data.source_slot]
+            || Some(vote.data.target) != state.historical_block_hashes[vote.data.target_slot]
+            || vote.data.target_slot <= vote.data.source_slot
+            || !is_justifiable_slot(&state.latest_finalized_slot, &vote.data.target_slot)
+        {
+            continue;
+        }
+
+        // Track attempts to justify new hashes
+        if !state.justifications.contains_key(&vote.data.target) {
+            let mut empty_justifications = Vec::<bool>::with_capacity(state.num_validators);
+            empty_justifications.resize(state.num_validators, false);
+
+            state
+                .justifications
+                .insert(vote.data.target, empty_justifications);
+        }
```

Comment:
```suggestion
        // Track attempts to justify new hashes
        if !state.justifications.contains_key(&vote.data.target) {
            state
                .justifications
                .insert(vote.data.target, vec![false; state.num_validators]);
        }
```

This can be simplified.